fix opening hours for other languages

Parthlakhani opened the pull request:
closes #10302 

Fix opening and closing hours for all languages

Parthlakhani committed:
GH-10302 fix opening hours for other languages

Parthlakhani committed:
Merge branch 'develop' into feature/GH-10302

Parthlakhani committed:
GH-10302 feat: Reset the store on language change

Parthlakhani committed:
feat: Add unit tests

Parthlakhani committed:
Merge branch 'develop' into feature/GH-10302

tobi-or-not-tobi commented on feature-libs/storefinder/components/store-finder-grid/store-finder-grid.component.ts:
```diff
@@ -1,16 +1,17 @@
 
 @Component({
   selector: 'cx-store-finder-grid',
   templateUrl: './store-finder-grid.component.html',
```
while you're at it, could you add the on push change detection?

tobi-or-not-tobi replied:
I believe this is still missing

tobi-or-not-tobi commented on feature-libs/storefinder/core/store/selectors/find-stores.selectors.ts:
```diff
@@ -28,3 +28,10 @@ export const getStoresLoading: MemoizedSelector<
   StateUtils.loaderLoadingSelector(state)
 );
+
+export const getStoresSucess: MemoizedSelector<
```
typo. 

if you have not already, you might want to use https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker

tobi-or-not-tobi commented on feature-libs/storefinder/components/store-finder-store/store-finder-store.component.ts:
```diff
@@ -23,11 +24,19 @@ export class StoreFinderStoreComponent implements OnInit {
     private routingService: RoutingService
   ) {}
 
+  loadData: Subscription = combineLatest([
```
I'd prefer to keep the component as clean as possible; we like to have the load/reload logic inside the facade layer or even lower if possible. You might want to look at other services where we've done similar things.

tobi-or-not-tobi commented on feature-libs/storefinder/components/store-finder-grid/store-finder-grid.component.ts:
```diff
@@ -20,11 +21,24 @@ export class StoreFinderGridComponent implements OnInit, OnDestroy {
+    FindStoresState
+  > = this.storeFinderService.getFindStoresEntities();
+  isLoaded$: Observable<boolean> = this.storeFinderService.getStoresLoaded();
+  loadData: Subscription = combineLatest([
```
See my note regarding clean components at store-finder-store.component.ts

Parthlakhani committed:
GH-10302 Move reload logic to facade

tobi-or-not-tobi commented on feature-libs/storefinder/components/store-finder-grid/store-finder-grid.component.ts:
```diff
@@ -19,12 +18,19 @@ export class StoreFinderGridComponent implements OnInit, OnDestroy {
+    this.findStores();
+  }
 
+  findStores() {
```
can you make this non-public, protected preferred.

tobi-or-not-tobi commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -28,7 +29,34 @@ export class StoreFinderService {
     protected routingService: RoutingService
-  ) {}
+  ) {
+    combineLatest([
```
We do not like to implement logic inside the constructor, as it blocks customisations in case one would extend this class. You could move the logic to a protected method, and call it from the constructor. The method can be further customised if needed.

tobi-or-not-tobi commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -28,7 +29,34 @@ export class StoreFinderService {
+      this.getStoresLoaded(),
+      this.getFindStoresEntities(),
+      this.routingService.getParams(),
+    ]).subscribe(([loading, loaded, value, params]) => {
```
if a subscribe is needed we must clean up the subscription, this is mainly because of leaking memory in an SSR process. You can implement the onDestroy lifecycle hook in the service. 

Ideally we don't have services that require explicit subscriptions. In this specific example it means that we're subscribing as soon as the service got initialised. That might not be the time that it's actually needed. We rather subscribe when the component explicitly requires this. You might be able to operate on the `getFindStoresEntities` observable that is returned to achieve the same. Tbh i've not looked into the details here.

Parthlakhani committed:
Merge branch 'develop' into feature/GH-10302

Parthlakhani committed:
feat: Call subscription when storeEntities are empty

Subscription in store-finder.service is called on language/curreny change only instead of service initialization.
Fixed logic for calling renderMap.
Added changeDetection and changed findStores to protected in store-finder-grid.copmonent

Parthlakhani committed:
Merge branch 'develop' into feature/GH-10302

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -186,4 +165,47 @@ export class StoreFinderService {
+              params.country &&
+              !params.store
+            ) {
+              this.findStoresAction(
```
You could create a function for this since it's the same as what is in the grid component.

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -186,4 +165,47 @@ export class StoreFinderService {
+        .subscribe(([value, loading, loaded, params]) => {
+          if (!loading && !loaded) {
+            if (
+              this.isEmpty(value.findStoresEntities) &&
```
Do you need this since you filter at the top?

Parthlakhani replied:
no, I can remove it.

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -186,4 +165,47 @@ export class StoreFinderService {
+            this.routingService.getParams()
+          )
+        )
+        .subscribe(([value, loading, loaded, params]) => {
```
I'd rename `value` as it's not very descriptive. Maybe also `params`.

Parthlakhani replied:
`Value` is not needed anymore since I removed the condition.

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -186,4 +165,47 @@ export class StoreFinderService {
     );
   }
+
+  protected initialize() {
```
I would rename this so it's more descriptive and add a JS-doc comment to describe what the function does.

LTiger14 commented on feature-libs/storefinder/components/store-finder-grid/store-finder-grid.component.ts:
```diff
@@ -30,7 +31,7 @@ export class StoreFinderGridComponent implements OnInit {
   }
 
-  findStores() {
+  protected findStores() {
```
While this is good, I wouldn't do it because of the breaking change.

Parthlakhani replied:
as discussed, leaving this `protected` as the method is new.

LTiger14 commented on feature-libs/storefinder/components/store-finder-grid/store-finder-grid.component.spec.ts:
```diff
@@ -62,14 +77,16 @@ describe('StoreFinderGridComponent', () => {
   beforeEach(() => {
     fixture = TestBed.createComponent(StoreFinderGridComponent);
     component = fixture.componentInstance;
+    StoreFinderGridComponent.prototype.ngOnInit = () => {};
```
Why are you doing this?

Parthlakhani replied:
It was for the test case `should call findStores on language change`, to avoid calling the method on `ngOnInit`
I will remove this.

LTiger14 commented on feature-libs/storefinder/components/store-finder-grid/store-finder-grid.component.spec.ts:
```diff
@@ -90,4 +107,13 @@ describe('StoreFinderGridComponent', () => {
     expect(component).toBeTruthy();
   });
+
+  it('should call findStores on language change', () => {
```
This component is not responsible for this any more right?

Parthlakhani replied:
yes, I will remove it.

LTiger14 commented on feature-libs/storefinder/components/store-finder-map/store-finder-map.component.ts:
```diff
@@ -25,7 +25,14 @@ export class StoreFinderMapComponent implements OnChanges {
   ngOnChanges(changes: SimpleChanges): void {
-    if (changes.locations && this.locations) {
+    if (
+      Object.entries(
```
Couldn't you add a validation in the store-data service? I think it would be simpler.

LTiger14 commented on feature-libs/storefinder/core/store/reducers/view-all-stores.reducer.ts:
```diff
@@ -5,7 +5,7 @@ export const initialState: ViewAllStoresState = {
 };
 
-export function reducer(
+export function viewAllStoresReducer(
```
Please update the unit test with this change.

LTiger14 commented on feature-libs/storefinder/core/store/reducers/find-stores.reducer.ts:
```diff
@@ -3,19 +3,24 @@ import { FindStoresState } from '../store-finder-state';
 };
 
-export function reducer(
+export function findStoresReducer(
```
Please update the unit test for this reducer.

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -145,4 +159,53 @@ export class StoreFinderService {
     }
     this.store.dispatch(callbackAction);
   }
+
```
Please add a test case for these changes. (all of them 😉 )

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -16,19 +16,26 @@ import {
     protected globalMessageService: GlobalMessageService,
-    protected routingService: RoutingService
-  ) {}
+    protected routingService: RoutingService,
```
You will need to add a constructor deprecation for this.

Parthlakhani committed:
Merge branch 'develop' into feature/GH-10302

Parthlakhani committed:
test: add unit tests for store finder service (GH-10302)

Added new test cases for reloadStoreEntitiesOnContextChange in store finder service.
Fixed broken unit tests.
Added a check in google-map-renderer service.
Added constructor deprecation in store finder service.
Removed unnecessary unit tests.

Parthlakhani committed:
Merge branch 'develop' into feature/GH-10302

LTiger14 commented on feature-libs/storefinder/components/store-finder-grid/store-finder-grid.component.ts:
```diff
@@ -1,41 +1,39 @@
+    this.findStores();
+  }
 
+  protected findStores() {
```
Forgot to mention it before. As a general rule always add an explicit return type, even if just `void` like this case.

LTiger14 commented on feature-libs/storefinder/components/store-finder-grid/store-finder-grid.component.spec.ts:
```diff
@@ -20,16 +20,22 @@ class MockStoreFinderListItemComponent {
-  getViewAllStoresEntities: jasmine.createSpy().and.returnValue(of(Observable)),
-  getViewAllStoresLoading: jasmine.createSpy(),
-  findStoresAction: jasmine.createSpy().and.returnValue(of(Observable)),
+  getFindStoresEntities: jasmine.createSpy().and.returnValue(of(Observable)),
```
Just to be cleaner you could do `import createSpy = jasmine.createSpy;` and just use `createSpy()`. You can check `user-details.component.spec.ts` to see what I mean.

LTiger14 commented on feature-libs/storefinder/components/store-finder-grid/store-finder-grid.component.spec.ts:
```diff
@@ -20,16 +20,22 @@ class MockStoreFinderListItemComponent {
   location;
 }
 
+class MockTranslationService {
```
Starting recently when creating a mock we want to use `implements Partial<...>` so that it's easier to make the changes when the base class is modified. You can cgeck this to see what I mean https://sap.github.io/spartacus-docs/type-safety-in-spartacus#partial-usage-for-service-or-component-mocks

LTiger14 commented on feature-libs/storefinder/components/store-finder-store/store-finder-store.component.spec.ts:
```diff
@@ -8,15 +8,22 @@ import {
-  viewStoreById: jasmine.createSpy().and.returnValue(of(Observable)),
-};
+const isLoading$: BehaviorSubject<Boolean> = new BehaviorSubject(true);
+const isLoaded$: BehaviorSubject<Boolean> = new BehaviorSubject(true);
```
If you don't actually do a `isLoaded$.next(...)` to test the behavior with a different value you don't need the `BehaviorSubject` you can just return `of(true)`.

Also in unit tests it's generally better to use `Subject` rather than `BehaviorSubject`. Basically `Subject` don't keep the value unless there is a subscription, this means you eliminate the risk for one test having an impact on the other.

Parthlakhani replied:
This was used by the previous test case that I removed, no need to use BehaviourSubject. I will remove it.

LTiger14 commented on feature-libs/storefinder/components/store-finder-store/store-finder-store.component.spec.ts:
```diff
@@ -8,15 +8,22 @@ import {
-};
+const isLoading$: BehaviorSubject<Boolean> = new BehaviorSubject(true);
+const isLoaded$: BehaviorSubject<Boolean> = new BehaviorSubject(true);
+class MockStoreFinderService {
```
Please use `Partial` (see store-finder-grid)

LTiger14 commented on feature-libs/storefinder/components/store-finder-stores-count/store-finder-stores-count.component.spec.ts:
```diff
@@ -4,10 +4,11 @@ import { I18nTestingModule } from '@spartacus/core';
 const mockStoreFinderService = {
   viewAllStores: jasmine.createSpy(),
-  getViewAllStoresEntities: jasmine.createSpy(),
+  getViewAllStoresEntities: jasmine.createSpy().and.returnValue(of(Observable)),
```
See `store-finder-grid`

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -145,4 +182,56 @@ export class StoreFinderService {
+    }
+  }
+
+  callFindStoresAction(routeParams) {
```
Please add type for parameter and return type.

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -145,4 +182,56 @@ export class StoreFinderService {
+  /**
+   * Reload store data when store entities are empty because of the context change
+   */
+  protected reloadStoreEntitiesOnContextChange() {
```
Please add return type.

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -145,4 +182,56 @@ export class StoreFinderService {
+   * Reload store data when store entities are empty because of the context change
+   */
+  protected reloadStoreEntitiesOnContextChange() {
+    if (isPlatformBrowser(this.platformId)) {
```
Since `platformId` is optional a customer extending this service might not have it provided. Could you format the if in a way it's true if -> the platform is browser (funciton return true) or if platformId is null/undefined.

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.ts:
```diff
@@ -145,4 +182,56 @@ export class StoreFinderService {
+              this.isEmpty(data.findStoresEntities) &&
+              this.isEmpty(data.findStoreEntitiesById)
+          ),
+          withLatestFrom(
```
This function is starting to look really good 👍 Could we just try this too see if it would work https://sap.github.io/spartacus-docs/events-and-subscriptions/ ? Basically moving the `withLatestFrom` to a `switchMap`.

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.spec.ts:
```diff
@@ -13,13 +19,21 @@ import {
+  [key: string]: string;
+}> = new BehaviorSubject({});
 
 class MockRoutingService {
```
While you are here you can add the partial 😉

LTiger14 commented on feature-libs/storefinder/core/facade/store-finder.service.spec.ts:
```diff
@@ -55,12 +70,41 @@ describe('StoreFinderService', () => {
+    mockStoreEntities
+  );
+
+  const mockSelect = (
```
Good stuff 👍

LTiger14 commented on feature-libs/storefinder/core/store/actions/view-all-stores.action.spec.ts:
```diff
@@ -40,3 +40,11 @@ describe('View All Stores Actions', () => {
     });
   });
 });
```
You can just put in the parent describe

LTiger14 commented on feature-libs/storefinder/core/store/selectors/find-stores.selectors.spec.ts:
```diff
@@ -56,4 +55,21 @@ describe('FindStores Selectors', () => {
     });
   });
+
+  describe('getStoresSucess', () => {
```
Small typo `getStoresSucess`

Parthlakhani committed:
refactor: Update unit tests

Added method and parameter types.
Updated declaration of mockService class.

Parthlakhani committed:
Merge branch 'develop' into feature/GH-10302

LTiger14 commented on package.json:
```diff
@@ -41,7 +41,7 @@
     "prettier": "prettier --config ./.prettierrc --list-different \"{projects,feature-libs,core-libs,integration-libs}/**/*{.ts,.js,.json,.scss,.html}\"",
     "prettier:fix": "prettier --config ./.prettierrc --list-different \"{projects,feature-libs,core-libs,integration-libs}/**/*{.ts,.js,.json,.scss,.html}\" --write",
-    "start": "env-cmd -e ci,b2c,$SPA_ENV ng serve",
+    "start": "env-cmd -e ci,b2b,$SPA_ENV ng serve",
```
why did you change this?

Parthlakhani replied:
Sorry, that was by mistake. I was running b2b for another ticket. I will change

Parthlakhani committed:
refactor: update package.json

LTiger14 commented on feature-libs/storefinder/core/store/store-finder-state.ts:
```diff
@@ -15,6 +15,7 @@ export interface StoresState {
 
 export interface FindStoresState {
   findStoresEntities: StoreEntities;
+  findStoreEntitiesById: StoreEntities;
```
Even though this is better but breaking changes let's not to it. Create a ticket to do it for 4.0.

LTiger14 replied:
Also remove all uses of it.

LTiger14 commented on feature-libs/storefinder/components/store-finder-grid/store-finder-grid.component.ts:
```diff
@@ -1,41 +1,39 @@
+    this.isLoading$ = this.storeFinderService.getStoresLoading();
+    this.locations$ = this.storeFinderService
+      .getFindStoresEntities()
+      .pipe(map((data) => data.findStoresEntities));
```
I'm adding it here because I can't place it in the service. 

- ALL of the methods that now need a map **should** return **exactly** the same thing that they did before. So add the `map` in there.
- Deprecate those methods for 4.0 -> say they will return the entities rather than data.
- Create new methods that do the same thing but are not called entities since technically they return the data.

Parthlakhani committed:
refactor: remove breaking changes (GH-10302)

closes #10302
Removed findStoreEntitiesById from storeFinderState to avoid breaking changes.
Moved map to service from component to avoid breaking change.

Parthlakhani committed:
refactor: remove blank line

Parthlakhani committed:
Merge branch 'develop' into feature/GH-10302

Parthlakhani committed:
refactor: change type of reducer
